### PR TITLE
CDAP-19777 make programId default version in different schedule services

### PR DESCRIPTION
**Pull request by jingtaowu3**
[CDAP-19777](https://cdap.atlassian.net/browse/CDAP-19777). ProgramId is very closely tied to Schedule, eg ProgramSchedule and ProgramStatusTrigger. Since Schedule is now versionless, it doesn't make sense to have a versioned ProgramId for those classes because the version can be changing. And when it comes to comparing two same programs with different version, the currently logic in scheduler will break.
Thus adding a new method to construct a ProgramId with default version out of an existing programId. This default version programId will be used/stored/compared for the schedule operations. 

As for backward compatibility, i assume once the LCM flag is turned on it is not allowed to turn off, so in that case ProgramId used for schedules should be "-SNAPSHOT" even before users upgrade their instances to 6.8, so this change should not break things during the upgrade.

### CODE REVIEW

File: cdap-proto/src/main/java/io/cdap/cdap/proto/id/ProgramId.java [hunk 1]
```diff
@@ -157,4 +146,12 @@ public Iterable<String> toIdParts() {
   public static ProgramId fromString(String string) {
     return EntityId.fromString(string, ProgramId.class);
   }
+
+  /**
+   * Schedule is versionless. Convert versionId in ProgramId to be default version, so it represents
+   * the latest version program
+   */
+  public ProgramId toDefaultVersion() {
```
Comment (sanjanasandeep) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/id/ProgramId.java]: is default : SNAPSHOT?

Comment (jingtaowu3) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/id/ProgramId.java]: yes

Comment (sanjanasandeep) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/id/ProgramId.java]: Do you think this solution is sustainable? what if we decide to map to latest in future instead of -SNAPSHOT, since everything as of today points to latest?

Comment (jingtaowu3) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/id/ProgramId.java]: For sustainable solution I think we probably want to bind `ProgramReference` with schedules, but that's not an easy change since ProgramId is already part of the table. This solution is also to accommodate existing schedules from 6.7 created with "-SNAPSHOT" version.  
Even today "-SNAPSHOT" is used to reference "latest" in other places like program LC handlers

Comment (sanjanasandeep) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/id/ProgramId.java]: I see your point, however my concern is that we are not being consistent with our approaches - like in some places we map to latest and some places (like here in schedules) we map to snapshot. Also, going forward we don't want to have this default snapshot I think, so we might run into loopholes. Maybe we should have a discussion on this after 6.8?  But yes, for now this solution works well.

**Review by sumengwang — APPROVED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java [hunk 2]
```diff
@@ -52,7 +52,7 @@ public ProgramSchedule(String name, String description,
                          ProgramId programId, Map<String, String> properties,
                          Trigger trigger, List<? extends Constraint> constraints, long timeoutMillis) {
     this.description = description;
-    this.programId = programId;
+    this.programId = programId.toDefaultVersion();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java]: Why is this needed? Isn't it easier to just ignore version throughout the scheduler code?

Comment (jingtaowu3) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java]: For `ProgramSchedule` this is not needed necessarily, during operations like listSchedule(programId), we will just need to compare all the other details expect versionId.  I was thinking making all programId "versionless" so we could do the comparison more easily.
However, this doesn't work for `ProgramStatusTrigger` which is another table - `triggerStore`. Because it has this string triggerkey concept that has programId in it (for example `programStatus:program:default.FakeApp.-SNAPSHOT.workflow.DataPipelineWorkflow.completed`), this trigger key is used to scan the table to find out which if there is any pipeline needs to be triggered on completion of `FakeApp.DataPipelineWorkflow`.  Here we cannot ignore the versionId without impacting query performance as it is part of the string,  but I am happy to know if there is a better workaround.

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java]: the trigger key is a detail internal to the store (which class is the triggerStore?). We can always insert a constant `-SNAPSHOT` as the version just in that store, instead of modifying all the callers to pass in a ProgramId with `-SNAPSHOT` version.

That being said, it's much cleaner to replace all the ProgramIds in the trigger/schedule classes with ProgramReference, is that happening in a separate PR?

Comment (jingtaowu3) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java]: good idea to just insert that constant in store, will try it out.
Yes replacing programId with ProgramReference will come in another PR

Comment (jingtaowu3) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java]: triggerStore belongs to ProgramScheduleStore

Comment (jingtaowu3) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java]: @albertshau I feel this is pretty much the minimum changes that can be made for schedule/trigger to work. Always inserting "-SNAPSHOT" resolves the triggering key prefix issue, but in a few places where it is comparing the `ProgramId` object directly, we had to make the version consistent.  Do you think we can go with this and replace `programId` with `ProgramReference` in a following PR (after https://github.com/cdapio/cdap/pull/14671 is merged)?

**Review by de-lan — APPROVED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/AbstractTimeSchedulerService.java [hunk 3]
```diff
@@ -135,7 +136,7 @@ public static String getTriggerName(ProgramId program, SchedulableProgramType pr
   }
 
   public static String programIdFor(ProgramId program, SchedulableProgramType programType) {
-    return String.format("%s:%s:%s:%s:%s", program.getNamespace(), program.getApplication(), program.getVersion(),
-                         programType.name(), program.getProgram());
+    return String.format("%s:%s:%s:%s:%s", program.getNamespace(), program.getApplication(),
+                         ApplicationId.DEFAULT_VERSION, programType.name(), program.getProgram());
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/AbstractTimeSchedulerService.java:140]: I don't think we actually want to do this and the programId.toDefaultVersion() method. Let's change things to use the ProgramReference and ApplicationReference classes from the other PR. If it's not going to get merged soon, I suggest that we pull those classes out into a separate PR so that they can be used by multiple people.

Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/AbstractTimeSchedulerService.java:140]: @albertshau Do you think we https://github.com/cdapio/cdap/pull/14671 is good enough to be approved/merged?

Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/AbstractTimeSchedulerService.java:140]: Otherwise I can open a separate PR introducing `ProgramReference, ApplicationReference, ProgramRunReference`

**Review by albertshau — APPROVED** [review]
